node_modules/col-lti/lib/api.js

```diff
@@ -0,0 +1,196 @@
+    if (err) {
+      log.error({'err': err}, 'Unable to query the database for the lti credentials');
+      return callback({'code': 500, 'msg': 'Unable to query the database for the lti credentials'});
+    } else if (!canvas) {
+      log.error({'err': err}, 'No credentials exist for the given api_domain and consumer_key');
+      return callback({'code': 400, 'msg': 'No credentials exist for the given api_domain and consumer_key'});
+    }
+
+    // Validate the LTI keys
+    var provider = new lti.Provider(consumer_key, canvas.lti_secret);
+    provider.valid_request(req, function(err, isValid) {
+      if (err) {
+        log.error({'err': err}, 'An LTI launch resulted in an error');
+        return callback({'code': 400, 'msg': err.message});
+      } else if (!isValid) {
+        log.warn('An LTI launch was invalid');
+        return callback({'code': 400, 'msg': 'Failed validation'});
+      }
+
+      // Create the course on the fly
+      options = {
+        'where': {
+          'canvas_course_id': canvas_course_id
+        },
+        'defaults': {
+          'canvas_canvas_api_domain': canvas.canvas_api_domain
+        }
+      };
+      DB.Course.findOrCreate(options).complete(function(err, data) {
+        if (err) {
+          log.error({'err': err}, 'Unable to query the database for the course');
+          return callback({'code': 500, 'msg': 'Unable to query the database for the course'});
+        }
+
+        var course = data[0];
+        var wasCreated = data[1];
+        if (wasCreated) {
+          log.info({'id': course.id}, 'Created a new course');
+        }
+
+        // Create the user on the fly
+        options = {
+          'where': {
+            'canvas_user_id': canvas_user_id,
+            'course_id': course.id
+          },
+          'defaults': {
+            'canvas_course_role': canvas_course_role,
+            'canvas_full_name': canvas_full_name,
+            'canvas_image': canvas_image,
+            'course_id': course.id,
+            'bookmarklet_token': CollabosphereUtil.generateString(32)
+          }
+        };
+        DB.User.findOrCreate(options).complete(function(err, data) {
+          if (err) {
+            log.error({'err': err}, 'Unable to query the database for the user');
+            return callback({'code': 400, 'msg': 'Unable to query the database for the user'});
+          }
+
+          var user = data[0];
+          var wasCreated = data[1];
+          if (wasCreated) {
+            log.info({'id': user.id}, 'Created a new user');
+          }
+
+          return callback(null, provider.body, user);
+        });
+      });
+    });
+  });
+};
+
+/**
+ * Get the Asset Library LTI Cartridge
+ *
+ * @param  {String}   host    The host for which to get the cartridge
+ * @return {String}           The Basic LTI Cartridge for the Asset Library
+ */
+var getAssetLibraryCartridge = module.exports.getAssetLibraryCartridge = function(host) {
```

Comment:
I have HTML pages sitting for all of the tools. It would be kind of nice if I could add them all to a course for UI development.